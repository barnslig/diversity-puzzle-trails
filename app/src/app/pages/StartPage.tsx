import {
  AppBar,
  Box,
  Button,
  Container,
  makeStyles,
  TextField,
  Toolbar,
  Typography,
} from "@material-ui/core";
import { FormattedMessage, useIntl } from "react-intl";
import { useFormik } from "formik";
import * as React from "react";

import useJoinGame from "../../common/hooks/useJoinGame";

const useStyles = makeStyles(() => ({
  appBarTitle: {
    flexGrow: 1,
  },
}));

type StartPageProps = {
  gameId?: string;
};

/**
 * Start Code Page to initially set the game ID
 *
 * The game ID is either supplied via the URL or collected from the user using a form.
 */
const StartPage = ({ gameId }: StartPageProps) => {
  const classes = useStyles();

  const intl = useIntl();
  const joinGame = useJoinGame();

  React.useEffect(() => {
    if (gameId) {
      // Automatically join the game when a game id is supplied via the url
      joinGame(gameId);
    }

    // This hook should only be executed once during mount, thus empty deps
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const formik = useFormik({
    initialValues: {
      gameId: gameId || "",
    },
    onSubmit: (values) => {
      joinGame(values.gameId);
    },
  });

  return gameId ? null : (
    <div>
      <AppBar position="fixed" color="inherit">
        <Toolbar>
          <Typography
            className={classes.appBarTitle}
            component="h1"
            variant="h6"
          >
            <FormattedMessage
              defaultMessage="An einem Spiel teilnehmen"
              description="start code page title"
            />
          </Typography>
        </Toolbar>
      </AppBar>
      <Box component="main" paddingTop={10}>
        <Container maxWidth="sm">
          <form onSubmit={formik.handleSubmit}>
            <Box marginBottom={2}>
              <Typography variant="body1">
                <FormattedMessage
                  defaultMessage="Um an einem Spiel teilzunehmen ist ein Spiel-Code erforderlich. Bitte tippe diesen hier ein:"
                  description="start code page description"
                />
              </Typography>
            </Box>

            <Box marginBottom={3}>
              <TextField
                label={intl.formatMessage({
                  defaultMessage: "Spiel-Code",
                  description: "start code page game id input label",
                })}
                variant="outlined"
                fullWidth
                required
                id="gameId"
                name="gameId"
                value={formik.values.gameId}
                onChange={formik.handleChange}
                error={formik.touched.gameId && Boolean(formik.errors.gameId)}
                helperText={formik.touched.gameId && formik.errors.gameId}
              />
            </Box>

            <Box display="flex" justifyContent="flex-end">
              <Button
                color="primary"
                fullWidth
                size="large"
                type="submit"
                variant="contained"
              >
                <FormattedMessage
                  defaultMessage="Am Spiel teilnehmen"
                  description="start code page submit button"
                />
              </Button>
            </Box>
          </form>
        </Container>
      </Box>
    </div>
  );
};

export default StartPage;
