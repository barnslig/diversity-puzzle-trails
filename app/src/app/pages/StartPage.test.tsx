import { screen, waitFor } from "@testing-library/react";
import * as React from "react";
import userEvent from "@testing-library/user-event";

import render from "../../common/testing/render";

import StartPage from "./StartPage";

const useJoinGame = jest.fn();
jest.mock("../../common/hooks/useJoinGame", () => ({
  __esModule: true,
  default: () => useJoinGame,
}));

it("renders nothing when a game id is supplied", () => {
  const { container } = render(<StartPage gameId="foobar" />);
  expect(container.firstChild).toBe(null);
});

it("sets the game id when a game id is supplied", async () => {
  render(<StartPage gameId="foobar" />);
  await waitFor(() => expect(useJoinGame).toHaveBeenCalledWith("foobar"));
});

it("renders correctly when no game id is supplied", () => {
  const { container } = render(<StartPage />);
  expect(container.firstChild).toMatchSnapshot();
});

it("lets the user input a game id when no game id is supplied", async () => {
  render(<StartPage />);

  userEvent.type(screen.getByLabelText(/Spiel-Code/i), "test-123");
  userEvent.click(screen.getByText(/Am Spiel teilnehmen/i));

  await waitFor(() => expect(useJoinGame).toHaveBeenCalledWith("test-123"));
});
