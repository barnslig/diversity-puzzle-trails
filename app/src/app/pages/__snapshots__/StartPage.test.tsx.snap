// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when no game id is supplied 1`] = `
<div
  class="MuiBox-root css-rrm59m"
>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-11b3ww9-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1t29gy6-MuiToolbar-root"
    >
      <h1
        class="MuiTypography-root MuiTypography-h6 css-1juivf6-MuiTypography-root"
      >
        An einem Spiel teilnehmen
      </h1>
    </div>
  </header>
  <main
    class="MuiBox-root css-130jr3z"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm css-cuefkz-MuiContainer-root"
    >
      <form
        action="#"
      >
        <div
          class="MuiBox-root css-1qm1lh"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Um an einem Spiel teilzunehmen ist ein Spiel-Code erforderlich. Bitte tippe diesen hier ein:
          </p>
        </div>
        <div
          class="MuiBox-root css-i3pbo"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="gameId"
              id="gameId-label"
            >
              Spiel-Code
              <span
                aria-hidden="true"
                class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="gameId"
                name="gameId"
                required=""
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-173wfxe"
                >
                  <span>
                    Spiel-Code
                     
                    *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-1bvc4cc"
        >
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth MuiButtonBase-root  css-189e9pj-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Am Spiel teilnehmen
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </form>
    </div>
  </main>
</div>
`;
